configfile: "config/config.yaml"


include: "rules/common.smk"
include: "rules/initial_setup.smk"
include: "rules/map_reads.smk"
include: "rules/process_bams.smk"
include: "rules/find_retrogenes.smk"
include: "rules/anchor_retrogenes.smk"
include: "rules/get_supporting_alignments.smk"
include: "rules/process_results.smk"
include: "rules/final_outputs.smk"


rule all:
    input:
        expand(
            "results/BAMS/{ref}/{sample}/transcriptome/short/{ref}.{sample}.transcriptome.short.sorted.bam",
            ref=refs,
            sample=short_samples,
        ),
        expand(
            "results/BAMS/{ref}/{sample}/transcriptome/long/{ref}.{sample}.transcriptome.long.sorted.bam",
            ref=refs,
            sample=long_samples,
        ),
        expand(
            "results/BAMS/{ref}/{sample}/genome/short/{ref}.{sample}.genome.short.sorted.bam.bai",
            ref=refs,
            sample=short_samples,
        ),
        expand(
            "results/BAMS/{ref}/{sample}/genome/long/{ref}.{sample}.genome.long.sorted.bam.bai",
            ref=refs,
            sample=long_samples,
        ),
        expand(
            f"results/Spanned/{{ref}}/{{sample}}/long/{{ref}}.{{sample}}.junctover{junction_overhang}.insertthresh{max_insertions}.longreads.spannedjunctions.tsv",
            ref=refs,
            sample=long_samples,
        ),
        expand(
            f"results/AS/{{ref}}/{{sample}}/{{ref}}.{{sample}}.junctover{junction_overhang}.insertthresh{max_insertions}.totalsupport{read_support}.genome.AS",
            ref=refs,
            sample=long_samples,
        ),
        expand(
            f"results/AS/{{ref}}/{{sample}}/{{ref}}.{{sample}}.junctover{junction_overhang}.insertthresh{max_insertions}.totalsupport{read_support}.genome.clustered.AS",
            ref=refs,
            sample=long_samples,
        ),
        expand(
            f"results/AS/{{ref}}/{{sample}}/{{ref}}.{{sample}}.junctover{junction_overhang}.insertthresh{max_insertions}.totalsupport{read_support}.genome.best.AS",
            ref=refs,
            sample=long_samples,
        ),
        expand(
            f"results/AS/{{ref}}/{{sample}}/{{ref}}.{{sample}}.junctover{junction_overhang}.insertthresh{max_insertions}.totalsupport{read_support}.genome.best.AS.diff",
            ref=refs,
            sample=long_samples,
        ),
        #expand(
        #    f"results/retrogenes/{{ref}}/{{sample}}/long/{{ref}}.{{sample}}.junctover{junction_overhang}.insertthresh{max_insertions}.totalsupport{read_support}.retrogenes.supportingalignments.genome.sorted.bam.bai",
        #    ref=refs,
        #    sample=long_samples,
        #),
        expand(
            f"results/retrogenes/{{ref}}/{{sample}}/long/{{ref}}.{{sample}}.junctover{junction_overhang}.insertthresh{max_insertions}.totalsupport{read_support}.retrogenes.final.results.tsv",
            ref=refs,
            sample=long_samples,
        ),
        expand(
            f"results/retrogenes/{{ref}}/{{sample}}/long/{{ref}}.{{sample}}.junctover{junction_overhang}.insertthresh{max_insertions}.totalsupport{read_support}.retrogenes.final.mainchrom.results.tsv",
            ref=refs,
            sample=long_samples,
        ),
        expand(
            f"results/retrogenes/{{ref}}/{{sample}}/long/{{ref}}.{{sample}}.junctover{junction_overhang}.insertthresh{max_insertions}.totalsupport{read_support}.retrogenes.chrommovement.results.tsv",
            ref=refs,
            sample=long_samples,
        ),
        expand(
            f"results/Spanned/{{ref}}/{{sample}}/short/{{ref}}.{{sample}}.junctover{junction_overhang}.insertthresh{max_insertions}.shortreads.spannedjunctions.tsv",
            ref=refs,
            sample=short_samples,
        ),
        expand(
            f"results/Summary/{{ref}}/{{sample}}/{{ref}}.{{sample}}.junctover{junction_overhang}.insertthresh{max_insertions}.totalsupport{read_support}.{strongthreshold}.results.txt",
            ref=refs,
            sample=short_samples,
        ),
        expand(
            f"results/Summary/{{ref}}/{{sample}}/{{ref}}.{{sample}}.junctover{junction_overhang}.insertthresh{max_insertions}.totalsupport{read_support}.{strongthreshold}.densityplots.png",
            ref=refs,
            sample=short_and_long_samples,
        ),
        expand(
            f"results/Summary/{{ref}}/{{sample}}/{{ref}}.{{sample}}.junctover{junction_overhang}.insertthresh{max_insertions}.totalsupport{read_support}.{strongthreshold}.upsetplots.1plus.tiff",
            ref=refs,
            sample=short_and_long_samples,
        ),
        expand(
            f"results/retrogenes/{{ref}}/{{sample}}/long/novel/{{ref}}.{{sample}}.junctover{junction_overhang}.insertthresh{max_insertions}.totalsupport{read_support}.novel",
            ref=refs,
            sample=long_samples,
        ),
        expand(
            "results/final_outputs/{ref}/{sample}/{sample}.longread.supportingalignments.genome.bam.bai",
            ref=refs,
            sample=long_samples,
        ),
        expand(
            "results/final_outputs/{ref}/{sample}/{sample}.longread.supportingalignments.transcriptome.bam.bai",
            ref=refs,
            sample=long_samples,
        ),
        expand(
            "results/final_outputs/{ref}/{sample}/{sample}.shortread.supportingalignments.transcriptome.bam.bai",
            ref=refs,
            sample=short_samples,
        ),
        expand(
            "results/final_outputs/{ref}/{sample}/{sample}.retrogene_consensus_sequences.fa",
            ref=refs,
            sample=long_samples,
        ),
        #expand(
        #    "results/final_outputs/{ref}/{sample}/{sample}.longread.retrogene_final_output.tsv",
        #    ref=refs,
        #    sample=long_samples,
        #),
        #expand(
        #    "results/final_outputs/{ref}/{sample}/{sample}.shortread.retrogene_final_output.tsv",
        #    ref=refs,
        #    sample=short_samples,
        #),
