<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Example Dataset &mdash; RetroDetector 0.1 documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script src="_static/doctools.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="FAQs" href="faqs.html" />
    <link rel="prev" title="Configuration" href="config.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="index.html" class="icon icon-home"> RetroDetector
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="start.html">Quick Start</a></li>
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation and Dependencies</a></li>
<li class="toctree-l1"><a class="reference internal" href="config.html">Configuration</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Example Dataset</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#drosophila-melanogaster">Drosophila melanogaster</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="faqs.html">FAQs</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">RetroDetector</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
      <li>Example Dataset</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/example.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="example-dataset">
<h1>Example Dataset<a class="headerlink" href="#example-dataset" title="Permalink to this heading"></a></h1>
<section id="drosophila-melanogaster">
<h2>Drosophila melanogaster<a class="headerlink" href="#drosophila-melanogaster" title="Permalink to this heading"></a></h2>
<p>In this example, we run the entire RetroDetector pipeline on a <em>Drosophila melanogaster</em> sample with Illumina paired-end and PacBio HiFi sequencing data. First make sure you have installed RetroDetector and all its depencies according to the <a class="reference internal" href="installation.html"><span class="doc">Installation and Dependencies</span></a> section. Then, make a working directory and change into that directory. Please make sure to run all of the following code while you are in your working directory. The first step is to download the necessary reference files from NCBI:</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">mkdir Downloads</span>
<span class="go">wget https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/001/215/GCF_000001215.4_Release_6_plus_ISO1_MT/GCF_000001215.4_Release_6_plus_ISO1_MT_genomic.fna.gz -O Downloads/ncbi_dmel.genomic.fna.gz</span>
<span class="go">wget https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/001/215/GCF_000001215.4_Release_6_plus_ISO1_MT/GCF_000001215.4_Release_6_plus_ISO1_MT_genomic.gtf.gz -O Downloads/ncbi_dmel.genomic.gtf.gz</span>
<span class="go">wget https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/001/215/GCF_000001215.4_Release_6_plus_ISO1_MT/GCF_000001215.4_Release_6_plus_ISO1_MT_rna_from_genomic.fna.gz -O Downloads/ncbi_dmel.rna_from_genomic.fna.gz</span>
</pre></div>
</div>
<p>The next step is to download and gzip the sequencing data. Make sure you have <a class="reference external" href="https://github.com/ncbi/sra-tools">sra-tools</a> installed. If you need more assistance please follow the instructions for <a class="reference external" href="https://github.com/ncbi/sra-tools/wiki/01.-Downloading-SRA-Toolkit">downloading</a>, <a class="reference external" href="https://github.com/ncbi/sra-tools/wiki/02.-Installing-SRA-Toolkit">installing</a>, and using <a class="reference external" href="https://github.com/ncbi/sra-tools/wiki/08.-prefetch-and-fasterq-dump">fasterq-dump</a>.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">prefetch SRR10238607 --max-size 21G</span>
<span class="go">fasterq-dump SRR10238607 --outdir Downloads</span>
<span class="go">gzip Downloads/SRR10238607.fastq</span>
<span class="go">prefetch SRR10728584</span>
<span class="go">fasterq-dump SRR10728584 --outdir Downloads</span>
<span class="go">gzip Downloads/SRR10728584_1.fastq</span>
<span class="go">gzip Downloads/SRR10728584_2.fastq</span>
</pre></div>
</div>
<p>The next step is to set up the configuration file config.yaml which tells RetroDetector where to find the necessary input files. If RetroDetector was installed at /user/software/RetroDetector, you would copy config.yaml to your working directory with the following command:</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">cp /users/software/RetroDetector/example/config.yaml .</span>
</pre></div>
</div>
<p>Now, you need edit config.yaml to add the scripts directory location from your RetroDetector installation. For example, if RetroDetector was installed at /user/software/RetroDetector, you would edit the scripts directory line of config.yaml as follows:</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">scripts_directory: /user/software/RetroDetector/scripts</span>
</pre></div>
</div>
<p>Next, copy Snakefile from your RetroDetector installation to your working directory. For example, if RetroDetector was installed at /user/software/RetroDetector and you are in your working directory, you would run:</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">cp /user/software/RetroDetector/Snakefile .</span>
</pre></div>
</div>
<p>Now you are ready to run RetroDetector. Activate your snakemake environment, and perform a dry run.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">conda activate snakemake</span>
<span class="go">snakemake -n</span>
</pre></div>
</div>
<p>If everything looks good, specify the number of available cores and run RetroDetector. For example, if your machine has 32 available cores, you would run:</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">snakemake --cores 32</span>
</pre></div>
</div>
<p>Once the pipeline finishes running, the following files should be in the RESULTS folder:</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">RESULTS</span>
<span class="go">├── genome.fa</span>
<span class="go">├── genome.fa.amb</span>
<span class="go">├── genome.fa.ann</span>
<span class="go">├── genome.fa.bwt</span>
<span class="go">├── genome.fa.fai</span>
<span class="go">├── genome.fa.pac</span>
<span class="go">├── genome.fa.sa</span>
<span class="go">└── samples</span>
<span class="go">    ├── A.fastq</span>
<span class="go">    ├── B.fastq</span>
<span class="go">    └── C.fastq</span>
</pre></div>
</div>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="config.html" class="btn btn-neutral float-left" title="Configuration" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="faqs.html" class="btn btn-neutral float-right" title="FAQs" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2022, T. Rhyker Ranallo-Benavidez, James E. Titus-McQuillan, Taylor D. Conway, Rebekah L. Rogers.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>